dll/win32/msafd/misc/dllmain.c

```diff
@@ -3490,30 +3619,35 @@ SetSocketInformation(PSOCKET_INFORMATION Socket,
                                    APCContext,
                                    IOSB,
                                    IOCTL_AFD_SET_INFO,
-                                   &InfoData,
-                                   sizeof(InfoData),
+                                   InfoData,
+                                   sizeof(AFD_INFO),
                                    NULL,
                                    0);
-
-    /* Wait for return */
-    if (Status == STATUS_PENDING && Overlapped == NULL)
+    if (Status == STATUS_PENDING)
     {
-        WaitForSingleObject(SockEvent, INFINITE);
+        if (!Overlapped)
+        {
+            /* Wait for completion of not overlapped */
+            WaitForSingleObject(SockEvent, INFINITE);
+        }
+
         Status = IOSB->Status;
     }
 
-    NtClose( SockEvent );
-
-    TRACE("Status %x Information %d\n", Status, IOSB->Information);
-
-    if (Status == STATUS_PENDING)
+    if (APCFunction)
     {
-        TRACE("Leaving (Pending)\n");
-        return WSA_IO_PENDING;
+        /* Allow APC to be processed */
+        NtSetEvent(SockEvent, NULL);
+    }
+    else
+    {
+        /* When using APC, this will be closed/freed by the APC function */
```

Comment:
It would be more logical to move this comment to the previous if-block, like:
`/* Cleanup will be done by the APC function */`,
while here, mention: `/* Not using APC, cleanup here */`

Reply:
Code changed.